### PR TITLE
PLANNER-1894 SolverManager.terminateEarly() fix

### PR DESCRIPTION
**Pull request by ge0ffrey**
https://issues.redhat.com/browse/PLANNER-1894

### CODE REVIEW

**Comment by ge0ffrey** [comment]
> Manual synchronization can, in this case, be replaced by replacing `SolverStatus` by `AtomicReference<SolverStatus>` and using `AtomicReference`'s `getAndSet(...)` methods. I believe that this mechanism is preferable in modern Java, as it will perform better and it makes it impossible for an unsynchronized operation to happen with `solverStatus` by accident.

I doubt it can be replaced, because the "NOT_SOLVING" change lock must include the map removal and the latch opening.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 1]
```diff
@@ -77,13 +78,19 @@ public ProblemId_ getProblemId() {
...
     @Override
     public Solution_ call() {
-        solverStatus = SolverStatus.SOLVING_ACTIVE;
+        synchronized (this) {
+            if (solverStatus != SolverStatus.SOLVING_SCHEDULED) {
+                // This job has been canceled before it started
+                return problemFinder.apply(problemId);
+            }
+            solverStatus = SolverStatus.SOLVING_ACTIVE;
+        }
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: @triceo I actually originally tried an AtomicRefence, for the reasons you mentioned.
Trouble was the code above versus the terminateEarly() code.
That latter did change drastically, so maybe there's a way to reintroduce the AtomicReference, let me try.

Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]:
```suggestion
            SolverStatus previousSolverStatus = solverStatus.getAndSet(SOLVING_ACTIVE);
            if (previousSolverStatus != SolverStatus.SOLVING_SCHEDULED) {
                // This job has been canceled before it started
                return problemFinder.apply(problemId);
            }
```

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 2]
```diff
@@ -96,12 +103,16 @@ public Solution_ call() {
...
-            solverStatus = SolverStatus.NOT_SOLVING;
-            terminatedLatch.countDown();
+            solvingTerminated();
         }
     }
 
+    private synchronized void solvingTerminated() {
+        solverManager.getProblemIdToSolverJobMap().remove(problemId);
+        solverStatus = SolverStatus.NOT_SOLVING;
+        terminatedLatch.countDown();
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]:
```suggestion
        SolverStatus previousSolverStatus = solverStatus.getAndSet(NOT_SOLVING);
        if (previousSolverStatus != NOT_SOLVING) {
            solverManager.getProblemIdToSolverJobMap().remove(problemId);
            terminatedLatch.countDown();
        }
```

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 3]
```diff
@@ -116,20 +127,26 @@ public Solution_ call() {
...
+            if (solverStatus == SolverStatus.SOLVING_SCHEDULED) {
+                solvingTerminated();
+            } else if (solverStatus == SolverStatus.SOLVING_ACTIVE) {
+                // Indirectly triggers solvingTerminated()
+                solver.terminateEarly();
+            } else if (solverStatus == SolverStatus.NOT_SOLVING) {
+                // Do nothing, solvingTerminated() already called
+            } else {
+                throw new IllegalStateException("Unsupported solverStatus (" + solverStatus + ").");
             }
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: This should be replacable by `switch (solverStatus.get())`, as all the actual state mutation happens in `solverTerminated()`/`solverTerminateEarly()` as I've suggested above.

**Comment by ge0ffrey** [comment]
@triceo Agreed, adjusting as you suggested. Good catch!

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 4]
```diff
@@ -47,8 +48,9 @@
     private final Consumer<? super Solution_> finalBestSolutionConsumer;
     private final BiConsumer<? super ProblemId_, ? super Throwable> exceptionHandler;
 
-    private volatile SolverStatus solverStatus;
+    private AtomicReference<SolverStatus> solverStatusReference;
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: Looks like this one could be `final`.

**Pull request merged**